### ISSUE TITLE
Feature request: customized function to open file

### ISSUE THREAD

**Issue by ipod825** [eugen0329/vim-esearch#93]
I like to use `tab drop` to keep one tab for only one file. Using `tabe` could results in too many duplicate tabs.

**Comment by eugen0329** [comment]
Probably, a good option would be provide an api to be able to open files using an arbitrary command. Of the top of my head:
```viml
call esearch#out#win#map('e', {-> b:esearch.open('tab drop')}) " runs :tab drop {filename}
```

Does that work for you?

**Comment by ipod825** [comment]
Yes. That works for me.

**Comment by ipod825** [comment]
This currently doesn't work in last commit of the development branch.

**Comment by eugen0329** [comment]
Can't reproduce. Maybe you set `g:esearch#env` to `'dev'`? It rereads files and probably resets some of your configs, that is convenient for development, but shouldn't happen on regular plugin usage. Also, provide an output from `map e` or another mapping that was broken, please.

**Comment by ipod825** [comment]
I found the mapping that cause this to fail:
`cnoremap <c-u> <c-u><BS>`

Do you have any idea on this?

**Comment by eugen0329** [comment]
Are you sure it's cnoremap? Documentation `:h map-overview` says it's for commandline.
If there are mappings you want to disable per buffer, then after #104 you could try something like `au User esearch#out#win#init_pre nunmap <buffer> <C-u>`

**Comment by eugen0329** [comment]
All the mappings are defined like `nnoremap  ... :<C-U>cal b:esearch.open('vnew')<CR>`. They contain `<C-u>`, but it's a generic pattern used by almost each plugin to define mappings and it should not expand `cnoremap <c-u> <c-u><BS>` inside. Also, I cannot reproduce the issue when add the cmap to my configs.

**Comment by eugen0329** [comment]
Also try to pull recent changes as they are exactly about opening api.

**Comment by ipod825** [comment]
> All the mappings are defined like nnoremap ... :<C-U>cal b:esearch.open('vnew')<CR>. They contain <C-u>, but it's a generic pattern used by almost each plugin to define mappings and it should not expand cnoremap <c-u> <c-u><BS> inside. 
 
I think this is the issue. I see no reason why `<c-u>` will not be expand. Using #104, I got `E31: No such mapping` for `:cunmap <buffer> <c-u>`. Only `:cunmap <c-u>` could work but discard the `<c-u>` mapping globally.

But as you mentioned that this is a general pattern adopted by other plugins, I guess I'll just deprecate this mapping. I've notice some problem caused by this mapping with vscode neovim plugin before.

For your reference. Here's the vim config file that can reproduce the issue:
```vim
let g:vim_dir=fnamemodify($MYVIMRC, ":h")."/"
" Download vim-plug
let vim_plug_file=expand(g:vim_dir . 'autoload/plug.vim')
if !filereadable(vim_plug_file)
    echo "Installing vim-plug.."
    echo ""
    execute 'silent !mkdir -p ' . g:vim_dir . 'autoload'
    call system('wget https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim -P ' . g:vim_dir . 'autoload')
endif

augroup POST_PLUG_END
    " This augroup enables us to call functions defined in a plugin before
    " plug#end() call.
    autocmd!
augroup END


call plug#begin(g:vim_dir . 'bundle')

Plug 'eugen0329/vim-esearch', {'branch': 'development'}
" vim-esearch {{{

augroup POST_PLUG_END
    autocmd USER PLUG_END call esearch#out#win#map('e', {-> b:esearch.open('tab drop')})
    autocmd USER esearch#out#win#init_pre cunmap <buffer> <C-u>
augroup END
" }}}

call plug#end()

silent doautocmd USER PLUG_END
cnoremap <c-u> <c-u><BS>
```

**Comment by eugen0329** [comment]
I've finally understood what was the problem. `<plug>...` sequence are defined without recursion using `cnoremap`, but must be referenced later with allowing it with `cmap e <plug>...`, while calling commandline directly without using them must always be defined without recursion. Another vim's quirk.

**Comment by eugen0329** [comment]
I think other plugins may have a similar issue, as you've mentioned. But it's really an edge case and at least after the latest dev branch commit everything should be fine.

**Comment by ipod825** [comment]
Nice. For your information. The command I use to open the searching result is `NewTabdrop` defined in this [plugin](https://github.com/ipod825/vim-tabdrop). It `tab drop` to a new tab while closing the esaerch buffer automatically.

**Comment by eugen0329** [comment]
To similarly maintain a single window for splits within a tab you may try {'once': 1} option and `b:esearch.opened_once_manager` to close or jumping to them. It's from latest vital, so you can reference their [docs](https://github.com/vim-jp/vital.vim/blob/master/doc/vital/Vim/BufferManager.txt) as well.

**Comment by ipod825** [comment]
Could you elaborate more? I found no `once` in either esearch document or vital's doc.

**Comment by eugen0329** [comment]
Vital provides only a useful backend to manipulate windows that is a real pain in vim. Docs will be later right before the merge when the API become more stable. Right now you can reference the [PR](https://github.com/eugen0329/vim-esearch/pull/104). There are several examples to understand how it works.

**Comment by eugen0329** [comment]
There's also a debouncing api you may be interested in. It works in the same way as it's done in js (lodash lib), so you can build your on workflow by defining a handler for cursor hold event using an arbitrary wait time (which will be independent of CursorHold, of course, thus no overwriting of a global `updatetime` that may be used by other plugins).

A common case is to define a preview or a split opener on hold to press less keys, but it also can be used, for example, to dig some information from a filename or a line (like feeding it to fugitive to fetch information about modified/deleted lines or something else). In other words, you are welcome to provide your ideas here on in the PR.
